File: docs/sphinx/examples/python/dynamics/cavity_qed.py
import cudaq
from cudaq.operator import *
import numpy as np
import cupy as cp
import os
import matplotlib.pyplot as plt

# Set the target to our dynamics simulator
cudaq.set_target("nvidia-dynamics")

# This example demonstrate a simulation of cavity quantum electrodynamics (interaction between light confined in a reflective cavity and atoms)

# System dimensions: atom (2-level system) and cavity (10-level system)
dimensions = {0: 2, 1: 10}

# Alias for commonly used operators
# Cavity operators
a = operators.annihilate(1)
a_dag = operators.create(1)

# Atom operators
sm = operators.annihilate(0)
sm_dag = operators.create(0)

# Defining the Hamiltonian for the system: self-energy terms and cavity-atom interaction term.
hamiltonian = 2 * np.pi * operators.number(1) + 2 * np.pi * operators.number(
    0) + 2 * np.pi * 0.25 * (sm * a_dag + sm_dag * a)

# Initial state of the system
# Atom in ground state
qubit_state = cp.array([[1.0, 0.0], [0.0, 0.0]], dtype=cp.complex128)

# Cavity in a state which has 5 photons initially
cavity_state = cp.zeros((10, 10), dtype=cp.complex128)
cavity_state[5][5] = 1.0
rho0 = cudaq.State.from_data(cp.kron(qubit_state, cavity_state))

steps = np.linspace(0, 10, 201)
schedule = Schedule(steps, ["time"])

# First, evolve the system without any collapse operators (ideal).
evolution_result = evolve(
    hamiltonian,
    dimensions,
    schedule,
    rho0,
    observables=[operators.number(1), operators.number(0)],
    collapse_operators=[],
    store_intermediate_results=True,
    integrator=ScipyZvodeIntegrator())

# Then, evolve the system with a collapse operator modeling cavity decay (leaking photons)
evolution_result_decay = evolve(
    hamiltonian,
    dimensions,
    schedule,
    rho0,
    observables=[operators.number(1), operators.number(0)],
    collapse_operators=[np.sqrt(0.1) * a],
    store_intermediate_results=True,
    integrator=ScipyZvodeIntegrator())

get_result = lambda idx, res: [
    exp_vals[idx].expectation() for exp_vals in res.expectation_values()
]
ideal_results = [
    get_result(0, evolution_result),
    get_result(1, evolution_result)
]
decay_results = [
    get_result(0, evolution_result_decay),
    get_result(1, evolution_result_decay)
]

fig = plt.figure(figsize=(18, 6))

plt.subplot(1, 2, 1)
plt.plot(steps, ideal_results[0])
plt.plot(steps, ideal_results[1])
plt.ylabel("Expectation value")
plt.xlabel("Time")
plt.legend(("Cavity Photon Number", "Atom Excitation Probability"))
plt.title("No decay")

plt.subplot(1, 2, 2)
plt.plot(steps, decay_results[0])
plt.plot(steps, decay_results[1])
plt.ylabel("Expectation value")
plt.xlabel("Time")
plt.legend(("Cavity Photon Number", "Atom Excitation Probability"))
plt.title("With decay")

abspath = os.path.abspath(__file__)
dname = os.path.dirname(abspath)
os.chdir(dname)
fig.savefig('cavity_qed.png', dpi=fig.dpi)

File: docs/sphinx/examples/cpp/dynamics/cavity_qed.cpp
/*******************************************************************************
 * Copyright (c) 2022 - 2025 NVIDIA Corporation & Affiliates.                  *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

#include "cudaq/algorithms/evolve.h"
#include "cudaq/dynamics_integrators.h"
#include "cudaq/base_integrator.h"
#include "cudaq/evolution.h"
#include "cudaq/operators.h"
#include "cudaq/schedule.h"
#include <cudaq.h>

int main() {

  // System dimensions:
  // subsystem 0 (atom) has 2 levels
  // subsystem 1 (cavity) has 10 levels
  std::map<int, int> dimensions{{0, 2}, {1, 10}};

  // For the cavity subsystem 1
  cudaq::product_operator<cudaq::matrix_operator> a = cudaq::boson_operator::annihilate(1);
  cudaq::product_operator<cudaq::matrix_operator> a_dag = cudaq::boson_operator::create(1);

  // For the atom subsystem 0
  cudaq::product_operator<cudaq::matrix_operator> sm = cudaq::boson_operator::annihilate(0);
  cudaq::product_operator<cudaq::matrix_operator> sm_dag = cudaq::boson_operator::create(0);

  cudaq::product_operator<cudaq::matrix_operator> atom_occ_op_t =
      cudaq::matrix_operator::number(0);
  cudaq::operator_sum<cudaq::matrix_operator> atom_occ_op(atom_occ_op_t);

  cudaq::product_operator<cudaq::matrix_operator> cavity_occ_op_t =
      cudaq::matrix_operator::number(1);
  cudaq::operator_sum<cudaq::matrix_operator> cavity_occ_op(cavity_occ_op_t);

  auto hamiltonian = (2 * M_PI * cavity_occ_op) + (2 * M_PI * atom_occ_op) + 
                     (2 * M_PI * 0.25 * (sm * a_dag + sm_dag * a));

  // Build the initial state
  // For the atom, the density matrix in the ground state
  std::vector<std::complex<double>> initial_state_vec(20, 0.0);
  initial_state_vec[10] = 1;

  // Define a time evolution schedule
  const int num_steps = 21;
  cudaq::Schedule schedule(cudaq::linspace(0.0, 10.0, num_steps));

  // Create a CUDA quantum state from a density matrix
  auto rho0 = cudaq::state::from_data(initial_state_vec);

  std::shared_ptr<cudaq::runge_kutta> integrator = std::make_shared<cudaq::runge_kutta>();
  integrator->dt = 0.001;
  integrator->order = 4;

  // Evolve without collapse operators
  cudaq::evolve_result evolve_result =
      cudaq::evolve(hamiltonian, dimensions, schedule, rho0, integrator, {},
        {cavity_occ_op, atom_occ_op}, true);

  std::shared_ptr<cudaq::runge_kutta> integrator_1 = std::make_shared<cudaq::runge_kutta>();
  integrator_1->dt = 0.001;
  integrator_1->order = 4;

  constexpr double decay_rate = 0.1;
  auto collapse_operator = std::sqrt(decay_rate) * a;
  // Evolve with collapse operators
  cudaq::evolve_result evolve_result_decay =
      cudaq::evolve(hamiltonian, dimensions, schedule, rho0, integrator_1,
                    {collapse_operator}, {cavity_occ_op, atom_occ_op}, true);

  // Lambda to extract expectation values for a given observable index
  auto get_expectation = [](int idx,
                            const auto &result) -> std::vector<double> {
    std::vector<double> expectations;

    auto all_exps = result.get_expectation_values().value();
    for (auto exp_vals : all_exps) {
      expectations.push_back((double)exp_vals[idx]);
    }
    return expectations;
  };

  auto ideal_result0 = get_expectation(0, evolve_result);
  auto ideal_result1 = get_expectation(1, evolve_result);
  auto decay_result0 = get_expectation(0, evolve_result_decay);
  auto decay_result1 = get_expectation(1, evolve_result_decay);

  std::cout << "Ideal result 0: ";
  for (auto val : ideal_result0) {
    std::cout << val << " ";
  }
  std::cout << std::endl;

  std::cout << "Ideal result 1: ";
  for (auto val : ideal_result1) {
    std::cout << val << " ";
  }
  std::cout << std::endl;

  std::cout << "Decay result 0: ";
  for (auto val : decay_result0) {
    std::cout << val << " ";
  }
  std::cout << std::endl;

  std::cout << "Decay result 1: ";
  for (auto val : decay_result1) {
    std::cout << val << " ";
  }
  std::cout << std::endl;

  std::cout << "Simulation complete. Plot saved to cavity_qed.png" << std::endl;
  return 0;
}